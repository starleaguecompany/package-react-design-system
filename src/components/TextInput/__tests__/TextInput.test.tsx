import * as React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Menu } from '@starleaguecompany/react-icons'
import { TextInput } from '../../../index'

describe('TextInput', () => {
  test('should render correctly with no props', () => {
    render(<TextInput />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('container')
  })

  test('should render correctly with attributes', () => {
    render(<TextInput id="test-id" className="test-class" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('test-class')
    expect(screen.getByRole('textbox')).toHaveAttribute('id', 'test-id')
  })

  test('should render correctly with value prop', () => {
    render(<TextInput value="Test text" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByRole('textbox')).toHaveValue('Test text')
    expect(screen.getByText('Test text')).toBeVisible()
  })

  test('should render correctly with disabled prop', () => {
    const { container } = render(<TextInput placeholder="Test placeholder" disabled />)
    const disabled = container.querySelectorAll('.disabled')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('disabled')
    expect(screen.getByText('Test placeholder')).toHaveClass('disabled')
    expect(disabled.length).toEqual(3)
    expect(screen.getByRole('textbox')).toBeDisabled()
  })

  test('should render correctly with invalid prop', () => {
    const { container } = render(<TextInput invalid />)
    const invalid = container.querySelectorAll('.invalid')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('invalid')
    expect(screen.getByRole('textbox')).toHaveClass('invalid')
    expect(invalid.length).toEqual(2)
  })

  test('should render correctly with loading prop', () => {
    const { container, rerender } = render(<TextInput loading />)
    const withIcon = container.querySelectorAll('.with-icon')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toContainElement(screen.getByTestId('Spinner'))
    expect(withIcon.length).toEqual(1)
    rerender(<TextInput />)
    expect(screen.getByTestId('TextInput')).not.toContainElement(screen.queryByTestId('Spinner'))
  })

  test('should render correctly with icon prop', () => {
    const { container } = render(<TextInput icon={<Menu />} />)
    const withIcon = container.querySelectorAll('.with-icon')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toContainElement(screen.getByTestId('Menu'))
    expect(withIcon.length).toEqual(1)
  })

  test('should render correctly with placeholder prop', () => {
    render(<TextInput placeholder="Test placeholder" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByText('Test placeholder')).toBeVisible()
    expect(screen.getByText('Test placeholder')).toHaveClass('placeholder')
  })

  test('should render correctly with postfix prop', () => {
    render(<TextInput value="100" postfix="₽" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveTextContent('100 ₽')
  })

  test('should render correctly with type prop', () => {
    render(<TextInput type="email" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByRole('textbox')).toHaveAttribute('type', 'email')
  })

  test('should render correctly with readOnly prop', () => {
    render(<TextInput readOnly value="Test value" />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByRole('textbox')).toHaveAttribute('readonly')
    userEvent.click(screen.getByRole('textbox'))
    userEvent.keyboard('New value')
    expect(screen.getByRole('textbox')).toHaveValue('Test value')
  })

  test('should execute onChange callback when typing', () => {
    const onChangeCallback = jest.fn()
    render(<TextInput value="Test value" onChange={onChangeCallback} />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    userEvent.click(screen.getByRole('textbox'))
    userEvent.keyboard('12345')
    expect(onChangeCallback).toHaveBeenCalledTimes(5)
  })

  test('should execute onFocus callback on click textInput', () => {
    const onFocusCallback = jest.fn()
    render(<TextInput onFocus={onFocusCallback} />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    userEvent.click(screen.getByRole('textbox'))
    expect(screen.getByRole('textbox')).toHaveFocus()
    expect(onFocusCallback).toHaveBeenCalledTimes(1)
  })

  test('should execute onBlur callback when blur focus', () => {
    const onBlurCallback = jest.fn()
    render(<TextInput onBlur={onBlurCallback} />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    userEvent.click(screen.getByRole('textbox'))
    userEvent.click(document.body)
    expect(screen.getByRole('textbox')).not.toHaveFocus()
    expect(onBlurCallback).toHaveBeenCalledTimes(1)
  })

  test('should render correctly with autoFocus prop', () => {
    const onFocusCallback = jest.fn()
    render(<TextInput autoFocus onFocus={onFocusCallback} />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('focused')
    expect(screen.getByRole('textbox')).toHaveFocus()
    expect(onFocusCallback).toHaveBeenCalledTimes(1)
  })

  test('should render correctly on external value update', () => {
    const onChangeCallback = jest.fn()
    const { rerender } = render(<TextInput value="123" onChange={onChangeCallback} />)

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    rerender(<TextInput value="1234" onChange={onChangeCallback} />)
    expect(screen.getByText('1234')).toBeVisible()
    expect(screen.getByRole('textbox')).toHaveValue('1234')
    expect(onChangeCallback).toBeCalledTimes(0)
  })

  test('should change disabled attribute', () => {
    const { container, rerender } = render(<TextInput placeholder="Test placeholder" />)
    let disabled = container.querySelectorAll('.disabled')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).not.toHaveClass('disabled')
    expect(screen.getByText('Test placeholder')).not.toHaveClass('disabled')
    expect(disabled.length).toEqual(0)
    expect(screen.getByRole('textbox')).not.toBeDisabled()

    rerender(<TextInput placeholder="Test placeholder" disabled />)
    disabled = container.querySelectorAll('.disabled')

    expect(screen.getByTestId('TextInput')).toMatchSnapshot()
    expect(screen.getByTestId('TextInput')).toHaveClass('disabled')
    expect(screen.getByText('Test placeholder')).toHaveClass('disabled')
    expect(disabled.length).toEqual(3)
    expect(screen.getByRole('textbox')).toBeDisabled()
  })
})
