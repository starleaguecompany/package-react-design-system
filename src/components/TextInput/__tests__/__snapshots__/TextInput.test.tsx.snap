// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should change disabled attribute 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <label
    class="placeholder"
  >
    Test placeholder
  </label>
</div>
`;

exports[`TextInput should change disabled attribute 2`] = `
<div
  class="container disabled"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input disabled"
        disabled=""
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <label
    class="placeholder disabled"
  >
    Test placeholder
  </label>
</div>
`;

exports[`TextInput should execute onBlur callback when blur focus 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should execute onChange callback when typing 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      >
        Test value
      </div>
      <input
        autocomplete="off"
        class="input filled"
        spellcheck="false"
        type="text"
        value="Test value"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should execute onFocus callback on click textInput 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly on external value update 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      >
        123
      </div>
      <input
        autocomplete="off"
        class="input filled"
        spellcheck="false"
        type="text"
        value="123"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with attributes 1`] = `
<div
  class="test-class container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        id="test-id"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with autoFocus prop 1`] = `
<div
  class="container focused active"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input active"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with disabled prop 1`] = `
<div
  class="container disabled"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input disabled"
        disabled=""
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <label
    class="placeholder disabled"
  >
    Test placeholder
  </label>
</div>
`;

exports[`TextInput should render correctly with icon prop 1`] = `
<div
  class="container with-icon"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <span
    class="icon container size-20 shape-round"
    data-qa="Icon"
  >
    <svg
      data-qa="Menu"
      fill="none"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 5h12M4 10h12M4 15h12"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </span>
</div>
`;

exports[`TextInput should render correctly with invalid prop 1`] = `
<div
  class="container invalid"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input invalid"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with loading prop 1`] = `
<div
  class="container with-icon"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <span
    class="icon container size-20"
    data-qa="Spinner"
  >
    <svg
      class="svg"
      viewBox="0 0 150 150"
    >
      <circle
        class="circle circle-1"
        cx="75"
        cy="75"
        r="60"
      />
      <circle
        class="circle circle-2"
        cx="75"
        cy="75"
        r="60"
      />
    </svg>
  </span>
</div>
`;

exports[`TextInput should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with placeholder prop 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
  <label
    class="placeholder"
  >
    Test placeholder
  </label>
</div>
`;

exports[`TextInput should render correctly with postfix prop 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      >
        100
      </div>
      <input
        autocomplete="off"
        class="input filled"
        spellcheck="false"
        type="text"
        value="100"
      />
      <div
        class="inputPostfix"
      >
         
        ₽
      </div>
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with readOnly prop 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      >
        Test value
      </div>
      <input
        autocomplete="off"
        class="input filled"
        readonly=""
        spellcheck="false"
        type="text"
        value="Test value"
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with type prop 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      />
      <input
        autocomplete="off"
        class="input"
        spellcheck="false"
        type="email"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextInput should render correctly with value prop 1`] = `
<div
  class="container"
  data-qa="TextInput"
>
  <div
    class="inputSegment centered"
  >
    <div
      class="inputWrapper"
    >
      <div
        class="inputValue"
      >
        Test text
      </div>
      <input
        autocomplete="off"
        class="input filled"
        spellcheck="false"
        type="text"
        value="Test text"
      />
    </div>
  </div>
</div>
`;
